### ISSUE TITLE
How to delete pose landmarks or add new pose landmarks?

### ISSUE THREAD

**Issue by Jakub112998** [google/mediapipe#2886]
The reason im writing is that Im trying to add new landmarks in torso, in between shoulders and in the middle of hips to connect them. At the same time I want to delete the unneccessary joints from face and thumbs.

I started my search in the pose.py and deleted 2 joins and added 3 joins:
```
class PoseLandmark(enum.IntEnum):
  NOSE = 0
  LEFT_EYE_INNER = 1
  LEFT_EYE = 2
  LEFT_EYE_OUTER = 3
  RIGHT_EYE_INNER = 4
  RIGHT_EYE = 5
  RIGHT_EYE_OUTER = 6
  LEFT_EAR = 7
  RIGHT_EAR = 8
  MOUTH_LEFT = 9
  MOUTH_RIGHT = 10
  LEFT_SHOULDER = 11
  RIGHT_SHOULDER = 12
  LEFT_ELBOW = 13
  RIGHT_ELBOW = 14
  LEFT_WRIST = 15
  RIGHT_WRIST = 16
  LEFT_PINKY = 17
  RIGHT_PINKY = 18
  LEFT_INDEX = 19
  RIGHT_INDEX = 20
  # LEFT_THUMB = 21 # these were deleted (not to draw nor save to csv)
  # RIGHT_THUMB = 22
  LEFT_HIP = 23
  RIGHT_HIP = 24
  LEFT_KNEE = 25
  RIGHT_KNEE = 26
  LEFT_ANKLE = 27
  RIGHT_ANKLE = 28
  LEFT_HEEL = 29
  RIGHT_HEEL = 30
  LEFT_FOOT_INDEX = 31
  RIGHT_FOOT_INDEX = 32
  # SHOULDERS_INDEX = 33 # these joints were added
  # TORSO_INDEX = 34
  # HIPS_INDEX = 35
```

in the pose_connections.py changed this part:
```
POSE_CONNECTIONS = frozenset([(0, 1), (1, 2), (2, 3), (3, 7), (0, 4), (4, 5),
                              (5, 6), (6, 8), (9, 10), (11, 12), (11, 13),
                              (13, 15), (15, 17), (15, 19), (15, 21), (17, 19),
                              (12, 14), (14, 16), (16, 18), (16, 20), (16, 22),
                              (18, 20), (11, 23), (12, 24), (23, 24), (23, 25),
                              (24, 26), (25, 27), (26, 28), (27, 29), (28, 30),
                              (29, 31), (30, 32), (27, 31), (28, 32), **(33, 34), 
                              (34, 35), (0, 33)**])
```

and in drawing_styles.py changed the code in lines 111-132 to (deleted left and right thumb):
# Pose
```
_THICKNESS_POSE_LANDMARKS = 2
_POSE_LANDMARKS_LEFT = frozenset([
    PoseLandmark.LEFT_EYE_INNER, PoseLandmark.LEFT_EYE,
    PoseLandmark.LEFT_EYE_OUTER, PoseLandmark.LEFT_EAR, PoseLandmark.MOUTH_LEFT,
    PoseLandmark.LEFT_SHOULDER, PoseLandmark.LEFT_ELBOW,
    PoseLandmark.LEFT_WRIST, PoseLandmark.LEFT_PINKY, PoseLandmark.LEFT_INDEX, 
    PoseLandmark.LEFT_HIP, PoseLandmark.LEFT_KNEE,
    PoseLandmark.LEFT_ANKLE, PoseLandmark.LEFT_HEEL,
    PoseLandmark.LEFT_FOOT_INDEX
])

_POSE_LANDMARKS_RIGHT = frozenset([
    PoseLandmark.RIGHT_EYE_INNER, PoseLandmark.RIGHT_EYE,
    PoseLandmark.RIGHT_EYE_OUTER, PoseLandmark.RIGHT_EAR,
    PoseLandmark.MOUTH_RIGHT, PoseLandmark.RIGHT_SHOULDER,
    PoseLandmark.RIGHT_ELBOW, PoseLandmark.RIGHT_WRIST,
    PoseLandmark.RIGHT_PINKY, PoseLandmark.RIGHT_INDEX, 
    PoseLandmark.RIGHT_HIP, PoseLandmark.RIGHT_KNEE,
    PoseLandmark.RIGHT_ANKLE, PoseLandmark.RIGHT_HEEL,
    PoseLandmark.RIGHT_FOOT_INDEX
])
```
and between 209-225:
 ```
 pose_landmark_style = {}
  left_spec = DrawingSpec(
      color=(0, 138, 255), thickness=_THICKNESS_POSE_LANDMARKS)
  right_spec = DrawingSpec(
      color=(231, 217, 0), thickness=_THICKNESS_POSE_LANDMARKS)
  for landmark in _POSE_LANDMARKS_LEFT:
    pose_landmark_style[landmark] = left_spec
  for landmark in _POSE_LANDMARKS_RIGHT:
    pose_landmark_style[landmark] = right_spec
  pose_landmark_style[PoseLandmark.NOSE] = DrawingSpec(
      color=_WHITE, thickness=_THICKNESS_POSE_LANDMARKS)
  pose_landmark_style[PoseLandmark.SHOULDERS_INDEX] = DrawingSpec( # want to add these (shoulders, torso, hips):
      color=_WHITE, thickness=_THICKNESS_POSE_LANDMARKS)
  pose_landmark_style[PoseLandmark.TORSO_INDEX] = DrawingSpec(
      color=_WHITE, thickness=_THICKNESS_POSE_LANDMARKS)
  pose_landmark_style[PoseLandmark.HIPS_INDEX] = DrawingSpec(
      color=_WHITE, thickness=_THICKNESS_POSE_LANDMARKS)
```

and after all these changes, this part - starting in bootstrap():
```
# Bootstrap all images.
      2 # Set limit to some small number for debug.
----> 3 bootstrap_helper.bootstrap(per_pose_class_limit=None)

<ipython-input-8-b13058f9160c> in bootstrap(self, per_pose_class_limit)
     99                  for lmk in pose_landmarks.landmark],
    100                 dtype=np.float32)
--> 101             assert pose_landmarks.shape == (31, 3),
```
is throwing and error, and then from there it is going to drawing_utils.py and throwing:
```
      if not (0 <= start_idx < num_landmarks and 0 <= end_idx < num_landmarks):
        raise ValueError(f'Landmark index is out of range. Invalid connection '
                         f'from landmark #{start_idx} to landmark #{end_idx}.')
```

so, I was looking for some hard coded value of 33 joints inside these files, yet didnt find any:

drawing_utils.py
pose_connections.py
pose.py
pose_test.py
drawing_styles.py 
solution_base.py
landmark_pb2.py

To sum up, my question is: How to not to draw some of the "default" joints and draw some additional ones.

Best,
J

**Comment by jiuqiant** [comment]
Instead of tweaking those data structures, you'd better to implement your own version of 

`def draw_landmarks(image: np.ndarray, landmark_list: landmark_pb2.NormalizedLandmarkList)` 

by following https://github.com/google/mediapipe/blob/master/mediapipe/python/solutions/drawing_utils.py#L120 and having the customizations.

I think the number 33 or 31 you mentioned is the length of `landmark_list`. Even it's not hardcoded, it's a fixed number returned by the solution output.

**Comment by Jakub112998** [comment]
@jiuqiant  is the landmark number fixed to 33 joints? We would like to add torso joint which is not possible in this model. Is there any way we can solve it?

Your help would be highly appreciated.

**Comment by jiuqiant** [comment]
> @jiuqiant is the landmark number fixed to 33 joints? We would like to add torso joint which is not possible in this model. Is there any way we can solve it?
> 
> Your help would be highly appreciated.

The solution API only outputs the landmarks that generated by the model. If you need additional joint, please calculate the joints based on the landmark output.

**Comment by il-nietos** [comment]
@Jakub112998 were you succesful in adding additional landmarks to the torso? This is exactly what I would like to do as well.

**Comment by jackson-sandland** [comment]
I'm also looking to do the same thing. Is this documented?

**Comment by knoguchi** [comment]
I don't think it's possible to add new landmarks unless Mediapipe train a new model with new joints.
